src/config/include/config/Node.hpp

```diff
@@ -0,0 +1,444 @@
+    }
+
+    template <
+        typename O,
+        typename... Args,
+        typename = std::enable_if_t<
+            !isNodeConstructible<O, Args...>() &&
+            std::is_same_v<decltype(YAML::convert<O>::encode(std::declval<O>())), YAML::Node>>,
+        typename = void,
+        typename = void>
+    std::optional<O> _maybeImpl(Args&&... args) const {
+        static_assert(sizeof...(args) == 0,
+                      "Cannot pass additional args when using built-in YAML conversion");
+        return std::make_optional<O>(_yaml.as<O>());
+    }
+
+    static YAML::Node parse(std::string);
+
+    template <typename K>
+    std::optional<const YAML::Node> yamlGet(const K& key) const {
+        if (!_valid) {
+            return std::nullopt;
+        }
+        const YAML::Node found = _yaml[key];
+        if (found) {
+            return std::make_optional<const YAML::Node>(found);
+        } else {
+            if (!_parent) {
+                return std::nullopt;
+            }
+            return _parent->yamlGet(key);
+        }
+    }
+
+    template <typename K>
+    const Node get(const K& key) const {
+        const std::string keyStr = v1::toString(key);
+        if (!_valid) {
+            return Node{YAML::Node{}, this, false, keyStr};
+        }
+        if constexpr (std::is_convertible_v<K, std::string>) {
+            if (key == "..") {
+                // this is...not the most succinct business-logic ever....
+                std::stringstream childKey;
+                if (_parent) {
+                    childKey << _parent->_key << "/";
+                }
+                childKey << _key << "/..";
+                if (!_parent) {
+                    return Node{YAML::Node{}, nullptr, false, childKey.str()};
+                }
+                return Node{_parent->_yaml, _parent->_parent, _parent->_valid, childKey.str()};
+            }
+        }
+        auto yaml = this->yamlGet(key);
+        if (yaml) {
+            return Node{*yaml, this, true, keyStr};
+        } else {
+            return Node{YAML::Node{}, this, false, keyStr};
+        }
+    }
+
+    void appendKey(std::ostringstream& out) const;
+
+    YAML::Node _yaml;
+    std::string _key;
+    const Node* _parent;
+    bool _valid;
+};
+
+
+// we can act like a Node if iterated value is a scalar or
+// we can act like a pair of Nodes if iterated value is a map entry
+class IteratorValue : public std::pair<Node, Node>, public Node {
+    using NodePair = std::pair<Node, Node>;
+
+public:
+    // jump through immense hoops to avoid knowing anything about the actual yaml iterator other
+    // than its pair form is a pair of {YAML::Node, YAML::Node}
+    template <typename ITVal>
+    IteratorValue(const Node* parent, ITVal itVal, size_t index)
+        : NodePair{std::make_pair(
+              Node{itVal.first,
+                   parent,
+                   itVal.first,
```

Comment:
Can `isNull()` replace `_valid`?

Reply:
There's a difference in `YAML::Node` between a null node and a node that doesn't exist:

```c++
    {
        YAML::Node yaml = YAML::Load("foo: null");

        REQUIRE(yaml["foo"].IsDefined() == true);
        REQUIRE(yaml["foo"].IsNull() == true);
        REQUIRE(bool(yaml["foo"]) == true);

        REQUIRE(yaml["bar"].IsDefined() == false);
        REQUIRE(yaml["bar"].IsNull() == false);
        REQUIRE(bool(yaml["bar"]) == false);
    }
```

Reply:
Derp. My question came out wrong.

⏪(rewind)

I had wanted to ask if _isValid is necessary or not.

I think whether a node exists or not is a YAML concept because a YAML file may or may not have a node; but for this genny::Node class, there should be a legitimate use case for every node that we create. If we don't expect a genny::Node to exist,  maybe() should be called instead of always getting back an empty Node.

There are two existing usages of `_valid`. One is in `operator bool()`, the other is in `get()`. Following my reasoning above, I think the former can be replaced with just `_yaml`, and the latter should throw an error and ask to either add a thing to the YAML or change the c++ call to user maybe().

Feel free to ignore the above if you disagree. I just wanted to see if we can simplify how we construct the iterator value. There might be legitimate use cases that need `_valid`.

Reply:
Oh I see - I like the intuition I didn't think about that at first.

Unfortunately it didn't pan out. I played with it and it made things worse. The problem is that we need a `YAML::Node _yaml` class-member, but in the cases where the path is invalid I just default-construct one, and unfortunately the default-constructed node doesn't behave the way you want or expect:

```c++
    {
        // This is why we need to have Node::_valid
        // We can't rely on our default-constructed nodes.
        auto defaultConstructed = YAML::Node{};
        REQUIRE(bool(defaultConstructed) == true);
        REQUIRE(defaultConstructed.Type() == YAML::NodeType::Null);
    }
```

I've added this to a test-case for posterity.

We could make the `YAML::Node` into a `std::optional` but that seems less clear than the `_valid` check

Reply:
It's weird to me that yaml-cpp has an internal notion of `ZombieNode` for this concept but doesn't let anyone use it.

https://github.com/jbeder/yaml-cpp/blob/yaml-cpp-0.6.2/include/yaml-cpp/node/impl.h#L18-L45